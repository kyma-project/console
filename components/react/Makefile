ci-pr: resolve build validate test
ci-master: resolve build validate test
ci-release: resolve build validate test

resolve:
	npm install

build:
	npm run build

validate:
	npm run lint-check
	eslint -c .eslinterrc ./src

test:
	CI=true npm run test
	