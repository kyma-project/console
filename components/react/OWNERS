reviewers:
    - default-reviewers

approvers:
    - default-approvers
    - Wawrzyn321
    - content-reviewers