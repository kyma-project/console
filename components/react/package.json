{
  "name": "@kyma-project/react-components",
  "version": "0.0.43",
  "main": "dist/index.js",
  "private": false,
  "license": "Apache-2.0",
  "directories": {
    "dist": "src"
  },
  "scripts": {
    "start:kyma": "npm run build:watch",
    "start:kyma:api": "npm run build:watch",
    "build": "webpack --config webpack.config.prod.js",
    "build:watch": "webpack -w --config webpack.config.dev.js",
    "test": "jest src/**; exit 0",
    "test:watch": "jest src/** --watch",
    "test:watchAll": "jest src/** --watchAll",
    "test:coverage:report": "open ./coverage/lcov-report/index.html",
    "lint-fix": "prettier --config ../../.prettierrc --write \"./src/**/*.{js,jsx,ts,tsx,html,css,json}\"",
    "lint-check": "prettier --config ../../.prettierrc --check \"./src/**/*.{js,jsx,ts,tsx,html,css,json}\""
  },
  "dependencies": {
    "@babel/runtime-corejs2": "^7.12.5",
    "lodash": "^4.17.21",
    "polished": "^1.8.1",
    "styled-components-modifiers": "^1.0.1"
  },
  "devDependencies": {
    "@babel/types": "7.0.0-beta.44",
    "ajv": "^6.12.3",
    "babel-cli": "^6.26.0",
    "babel-core": "^6.26.0",
    "babel-eslint": "^8.0.1",
    "babel-jest": "^22.4.1",
    "babel-loader": "^7.1.2",
    "babel-polyfill": "^6.26.0",
    "babel-preset-env": "^1.6.1",
    "babel-preset-react": "^6.24.1",
    "babel-preset-stage-0": "^6.24.1",
    "enzyme": "^3.1.0",
    "enzyme-adapter-react-16": "^1.0.2",
    "exec-sh": "^0.3.3",
    "file-loader": "^1.1.11",
    "fiori-fundamentals": "1.5.3",
    "fontfaceobserver": "^2.0.13",
    "handlebars": "^4.7.3",
    "jest": "^24.0.0",
    "jest-styled-components": "^4.11.0-1",
    "npm-run-all": "^4.1.1",
    "raf": "^3.4.0",
    "react-jsonschema-form": "1.7.0",
    "react-modal": "^3.5.1",
    "react-scrollbar": "^0.5.4",
    "react-test-renderer": "^16.0.0",
    "style-loader": "^0.20.2",
    "styled-components": "^3.2.3",
    "styled-components-grid": "^2.2.0",
    "webpack": "^4.28.1",
    "webpack-cli": "^3.2.1",
    "webpack-merge": "^4.2.1"
  },
  "peerDependencies": {
    "styled-components": "^3.2.3"
  },
  "resolutions": {
    "uglify-es": "3.3.7"
  },
  "jest": {
    "collectCoverage": true,
    "collectCoverageFrom": [
      "src/**",
      "!src/**/*.snap",
      "!src/components/index.js",
      "!src/modifiers/**",
      "!src/index.js"
    ],
    "moduleNameMapper": {
      "^.+\\.(css|less|scss)$": "babel-jest"
    },
    "moduleDirectories": [
      "node_modules",
      "./"
    ],
    "setupFiles": [
      "raf/polyfill",
      "./test_setup/config"
    ]
  }
}
