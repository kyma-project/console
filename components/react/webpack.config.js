const webpack = require('webpack');
const path = require('path');
const pkg = require('./package.json');

require('babel-polyfill');

let libraryName = pkg.name;

module.exports = {
  entry: ['./src/index'],
  module: {
    loaders: [
      { test: /\.js?$/, loader: 'babel-loader', exclude: /node_modules/ },
      {
        test: /\.css$/,
        use: [{ loader: 'style-loader' }, { loader: 'css-loader' }],
      },
      {
        test: /\.(eot|svg|ttf|woff|woff2|otf)$/,
        loader: 'file-loader?name=fonts/[name].[ext]',
      },
    ],
  },
  resolve: {
    modules: [path.resolve('./node_modules'), 'node_modules'],
    extensions: ['.js'],
  },
  output: {
    path: path.join(__dirname, '/lib'),
    publicPath: '/',
    filename: 'index.js',
    library: libraryName,
    libraryTarget: 'umd',
    umdNamedDefine: true,
  },
  devtool: 'source-map',
  devServer: {
    contentBase: './src',
    hot: true,
  },
  plugins: [
    new webpack.optimize.OccurrenceOrderPlugin(),
    new webpack.HotModuleReplacementPlugin(),
    new webpack.NoEmitOnErrorsPlugin(),
  ],
  externals: {
    'styled-components': {
      commonjs: 'styled-components',
      commonjs2: 'styled-components',
      amd: 'styled-components',
    },
  },
};
