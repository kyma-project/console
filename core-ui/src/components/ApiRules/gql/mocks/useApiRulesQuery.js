import { GET_API_RULES } from 'gql/queries';
import { API_RULE_EVENT_SUBSCRIPTION } from 'gql/subscriptions';
import { TESTING_STATE } from 'components/Lambdas/helpers/testing';

export const apiRuleMock = {
  name: 'apiRule',
  generation: 'generation',
  spec: {
    rules: [
      {
        path: '/.*',
        methods: ['GET', 'POST'],
        accessStrategies: [
          {
            handler: 'nope',
            config: {},
          },
        ],
      },
    ],
    service: {
      host: 'host',
      name: 'service',
      port: '80',
    },
  },
  status: {
    apiRuleStatus: {
      code: 'OK',
      description: '',
    },
  },
};

export const GET_API_RULES_ERROR_MOCK = variables => ({
  request: {
    query: GET_API_RULES,
    variables,
  },
  error: new Error(TESTING_STATE.ERROR),
});

export const GET_API_RULES_DATA_MOCK = (
  variables,
  APIRules = [apiRuleMock],
) => ({
  request: {
    query: GET_API_RULES,
    variables,
  },
  result: {
    data: {
      APIRules,
    },
  },
});

export const API_RULE_EVENT_SUBSCRIPTION_MOCK = (
  variables,
  apiRuleEvent = { type: 'ADD', apiRule: apiRuleMock },
) => ({
  request: {
    query: API_RULE_EVENT_SUBSCRIPTION,
    variables,
  },
  result: {
    data: {
      apiRuleEvent,
    },
  },
});
