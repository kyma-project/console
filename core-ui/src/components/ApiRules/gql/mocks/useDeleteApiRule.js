import { DELETE_API_RULE } from 'gql/mutations';
import { TESTING_STATE } from 'components/Lambdas/helpers/testing';

export const DELETE_API_RULE_ERROR_MOCK = variables => ({
  request: {
    query: DELETE_API_RULE,
    variables,
  },
  error: new Error(TESTING_STATE.ERROR),
});

export const DELETE_API_RULE_DATA_MOCK = variables => ({
  request: {
    query: DELETE_API_RULE,
    variables,
  },
  result: {
    data: {
      deleteAPIRule: {
        name: variables.name,
      },
    },
  },
});
