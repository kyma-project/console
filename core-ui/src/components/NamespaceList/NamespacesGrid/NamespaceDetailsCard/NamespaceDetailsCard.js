import React from 'react';
import LuigiClient from '@kyma-project/luigi-client';
import PropTypes from 'prop-types';
import classNames from 'classnames';

import { DELETE_NAMESPACE } from './../../../../gql/mutations';
import { useMutation } from '@apollo/react-hooks';

import { Button, Menu, Panel, Badge, Popover } from 'fundamental-react';
import './NamespaceDetailsCard.scss';

NamespaceDetailsCard.propTypes = {
  name: PropTypes.string.isRequired,
  allPodsCount: PropTypes.number.isRequired,
  healthyPodsCount: PropTypes.number.isRequired,
  status: PropTypes.string.isRequired,
  isSystemNamespace: PropTypes.bool.isRequired,
  applications: PropTypes.array,
};

export function getPodsRatioColor(healthyPods, allPods) {
  if (allPods === 0) {
    return '#107e3e';
  }

  const podsRatio = healthyPods / allPods;

  if (podsRatio === 1) {
    return '#107e3e';
  } else if (podsRatio > 0.8) {
    return '#e9730c';
  } else {
    return '#bb0000';
  }
}

function navigateToNamespaceDetails(namespaceName) {
  LuigiClient.linkManager().navigate(
    `/home/namespaces/${namespaceName}/details`,
  );
  LuigiClient.sendCustomMessage({ id: 'console.refreshNavigation' });
}

const Spinner = () => (
  <div className="fd-loading-dots" aria-hidden="false" aria-label="Loading">
    <div></div>
    <div></div>
    <div></div>
  </div>
);

export default function NamespaceDetailsCard({
  name,
  allPodsCount,
  healthyPodsCount,
  status,
  isSystemNamespace,
  applications,
}) {
  const [deleteNamespace] = useMutation(DELETE_NAMESPACE);

  const handleNamespaceDelete = () => {
    LuigiClient.uxManager()
      .showConfirmationModal({
        header: `Remove ${name}`,
        body: `Are you sure you want to delete namespace "${name}"?`,
        buttonConfirm: 'Delete',
        buttonDismiss: 'Cancel',
      })
      .then(async () => {
        try {
          await deleteNamespace({ variables: { name } });
        } catch (e) {
          console.warn(e);
          LuigiClient.uxManager().showAlert({
            text: e.message,
            type: 'error',
            closeAfter: 10000,
          });
        }
      })
      .catch(() => {});
  };

  const getPodsRatioStyle = () => {
    return { color: getPodsRatioColor(healthyPodsCount, allPodsCount) };
  };

  const popoverBody = (
    <Menu>
      <Menu.List>
        <Menu.Item
          onClick={e => {
            e.stopPropagation();
            handleNamespaceDelete(name);
          }}
        >
          Delete
        </Menu.Item>
      </Menu.List>
    </Menu>
  );

  const popoverControl = (
    <Button
      onClick={e => e.stopPropagation()}
      glyph="overflow"
      option="light"
      className="popover-control"
      aria-label="namespace-actions"
    />
  );

  const isTerminating = status === 'Terminating';
  const panelClass = classNames('namespace-details-card', {
    'namespace-details-card--terminating': isTerminating,
  });

  return (
    <Panel
      role="gridcell"
      className={panelClass}
      onClick={() => navigateToNamespaceDetails(name)}
    >
      <Panel.Header className="fd-has-color-text-1 fd-has-type-1">
        <Panel.Head aria-label="namespace-name" title={name} />
        <span>
          {isSystemNamespace && (
            <Badge className="fd-has-margin-left-tiny">System</Badge>
          )}
        </span>
        <Popover
          body={popoverBody}
          control={popoverControl}
          placement="right"
          noArrow
          disabled={isTerminating}
        />
      </Panel.Header>
      <Panel.Body>
        <section className="namespace-details-card__body">
          <div>
            <p
              className="fd-has-type-4 fd-has-font-weight-light"
              style={getPodsRatioStyle()}
            >
              {healthyPodsCount}/{allPodsCount}
            </p>
            <p>Pods are healthy</p>
          </div>
          {applications !== null && (
            <div>
              <p className="fd-has-type-4 fd-has-font-weight-light">
                {applications.length}
              </p>
              <p>
                {applications.length === 1
                  ? 'Bound Application'
                  : 'Bound Applications'}
              </p>
            </div>
          )}
        </section>
      </Panel.Body>
      {isTerminating && (
        <div className="overlay">
          <div className="overlay-caption fd-has-type-3">
            Terminating...
            <Spinner />
          </div>
        </div>
      )}
    </Panel>
  );
}
