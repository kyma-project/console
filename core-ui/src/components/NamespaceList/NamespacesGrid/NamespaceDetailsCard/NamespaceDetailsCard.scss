.namespace-details-card {
    position: relative;
    cursor: pointer;

    .fd-panel__header {
        display: flex;
        justify-content: space-between;
    }

    .popover-control::before {
        font-size: 16px;
    }
    
    .namespace-details-card__body {
        display: flex;
        justify-content: space-between;
        position: relative;
        width: 100%;
        height: 100%;
    }
}

.namespace-details-card--terminating {
    pointer-events: none;

    .overlay {
        pointer-events: none;
        // use hex, as variable won't work width rgba
        background-color: rgba(#D9D9D9, 0.3); // fd-color-neutral-3
        position: absolute;
        width: 100%;
        height: 100%;
        left: 0;
        top: 0;
    }
    
    .overlay-caption {
        display: block;
        position: absolute;
        left: 35%;
        top: 25%;
    }
}
