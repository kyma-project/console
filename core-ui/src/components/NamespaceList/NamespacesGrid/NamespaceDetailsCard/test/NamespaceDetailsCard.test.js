import React from 'react';
import { render, fireEvent, wait } from '@testing-library/react';
import { MockedProvider } from '@apollo/react-testing';

import NamespaceDetailsCard from './../NamespaceDetailsCard';
import { mockNavigate, deleteNamespaceMock } from './mocks';

jest.mock('@kyma-project/luigi-client', () => ({
  linkManager: () => ({
    navigate: mockNavigate,
  }),
  uxManager: () => ({
    addBackdrop: jest.fn(),
    removeBackdrop: jest.fn(),
    showConfirmationModal: () => Promise.resolve(),
  }),
  sendCustomMessage: jest.fn(),
}));

describe('NamespaceDetailsCard', () => {
  const applications = [{}, {}];
  it('Displays basic namespace data', () => {
    const { queryByText } = render(
      <NamespaceDetailsCard
        name="test-namespace-name"
        allPodsCount={10}
        healthyPodsCount={9}
        status="Active"
        isSystemNamespace={false}
        applications={applications}
      />,
    );
    expect(queryByText('9/10')).toBeInTheDocument();
    expect(queryByText(applications.length.toString())).toBeInTheDocument();
    expect(queryByText('Bound Applications')).toBeInTheDocument();
    expect(queryByText('test-namespace-name')).toBeInTheDocument();
    expect(queryByText('System')).not.toBeInTheDocument();
  });

  it(`Doesn't display applications if null`, () => {
    const { queryByText } = render(
      <NamespaceDetailsCard
        name="test-namespace-name"
        allPodsCount={10}
        healthyPodsCount={9}
        status="Active"
        isSystemNamespace={false}
        applications={null}
      />,
    );
    expect(queryByText('Bound Application')).not.toBeInTheDocument();
  });

  it('Displays "SYSTEM" badge on system namespace', () => {
    const { queryByText } = render(
      <NamespaceDetailsCard
        name="test"
        allPodsCount={10}
        healthyPodsCount={10}
        status="Active"
        isSystemNamespace={true}
        applications={applications}
      />,
    );

    expect(queryByText('System')).toBeInTheDocument();
  });

  it('Navigates to details on card click', async () => {
    const { getByRole } = render(
      <NamespaceDetailsCard
        name="test"
        allPodsCount={10}
        healthyPodsCount={10}
        status="Active"
        isSystemNamespace={true}
        applications={applications}
      />,
    );

    fireEvent.click(getByRole('gridcell'));

    await wait(() =>
      expect(mockNavigate).toHaveBeenCalledWith(
        '/home/namespaces/test/details',
      ),
    );
  });

  it('Triggers mutation on "Delete" click', async () => {
    //for Popovers's Warning: `NaN` is an invalid value for the `left` css style property.
    console.error = jest.fn();

    const { getByLabelText, getByText } = render(
      <MockedProvider mocks={[deleteNamespaceMock]} addTypename={false}>
        <NamespaceDetailsCard
          name="test"
          allPodsCount={10}
          healthyPodsCount={10}
          status="Active"
          isSystemNamespace={true}
          applications={applications}
        />
      </MockedProvider>,
    );

    // open popover
    fireEvent.click(getByLabelText('namespace-actions'));

    // delete namespace
    fireEvent.click(getByText('Delete'));

    await wait(() => expect(deleteNamespaceMock.result).toHaveBeenCalled());
  });
});
