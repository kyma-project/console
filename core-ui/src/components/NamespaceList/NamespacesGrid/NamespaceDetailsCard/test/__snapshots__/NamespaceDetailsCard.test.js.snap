// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NamespaceDetailsCard Displays "SYSTEM" badge 1`] = `
<div
  className="fd-panel namespace-details-card"
  onClick={[Function]}
>
  <div
    className="fd-panel__header fd-has-color-text-1 fd-has-type-1"
  >
    <span>
      test
      <span
        className="fd-badge fd-has-margin-left-tiny"
      >
        System
      </span>
    </span>
    <div
      className="fd-popover"
    >
      <div
        onFocus={[Function]}
        onMouseDown={[Function]}
        onTouchStart={[Function]}
      >
        <div
          className="fd-popover__control"
        >
          <button
            className="fd-button--light sap-icon--overflow popover-control"
            onClick={[Function]}
          />
        </div>
      </div>
    </div>
  </div>
  <div
    className="fd-panel__body"
  >
    <section
      className="namespace-details-card__body"
    >
      <div>
        <p
          className="fd-has-type-4 fd-has-font-weight-light"
          style={
            Object {
              "color": "#107e3e",
            }
          }
        >
          10
          /
          10
        </p>
        <p>
          Pods are healthy
        </p>
      </div>
      <div>
        <p
          className="fd-has-type-4 fd-has-font-weight-light"
        >
          0
        </p>
        <p>
          Bound Applications
        </p>
      </div>
    </section>
  </div>
</div>
`;

exports[`NamespaceDetailsCard Displays overlay when status is "Terminating" 1`] = `
<div
  className="fd-panel namespace-details-card namespace-details-card--terminating"
  onClick={[Function]}
>
  <div
    className="fd-panel__header fd-has-color-text-1 fd-has-type-1"
  >
    <span>
      test
    </span>
    <div
      className="fd-popover"
    >
      <div
        onFocus={[Function]}
        onMouseDown={[Function]}
        onTouchStart={[Function]}
      >
        <div
          className="fd-popover__control"
        >
          <button
            className="fd-button--light sap-icon--overflow popover-control"
            onClick={[Function]}
          />
        </div>
      </div>
    </div>
  </div>
  <div
    className="fd-panel__body"
  >
    <section
      className="namespace-details-card__body"
    >
      <div>
        <p
          className="fd-has-type-4 fd-has-font-weight-light"
          style={
            Object {
              "color": "#107e3e",
            }
          }
        >
          10
          /
          10
        </p>
        <p>
          Pods are healthy
        </p>
      </div>
      <div>
        <p
          className="fd-has-type-4 fd-has-font-weight-light"
        >
          0
        </p>
        <p>
          Bound Applications
        </p>
      </div>
    </section>
  </div>
  <div
    className="overlay"
  >
    <div
      className="overlay-caption fd-has-type-3"
    >
      Terminating...
      <div
        aria-hidden="false"
        aria-label="Loading"
        class="fd-loading-dots"
      >
        <div />
        <div />
        <div />
      </div>
    </div>
  </div>
</div>
`;

exports[`NamespaceDetailsCard Renders with minimal props 1`] = `
<div
  className="fd-panel namespace-details-card"
  onClick={[Function]}
>
  <div
    className="fd-panel__header fd-has-color-text-1 fd-has-type-1"
  >
    <span>
      test
    </span>
    <div
      className="fd-popover"
    >
      <div
        onFocus={[Function]}
        onMouseDown={[Function]}
        onTouchStart={[Function]}
      >
        <div
          className="fd-popover__control"
        >
          <button
            className="fd-button--light sap-icon--overflow popover-control"
            onClick={[Function]}
          />
        </div>
      </div>
    </div>
  </div>
  <div
    className="fd-panel__body"
  >
    <section
      className="namespace-details-card__body"
    >
      <div>
        <p
          className="fd-has-type-4 fd-has-font-weight-light"
          style={
            Object {
              "color": "#107e3e",
            }
          }
        >
          10
          /
          10
        </p>
        <p>
          Pods are healthy
        </p>
      </div>
      <div>
        <p
          className="fd-has-type-4 fd-has-font-weight-light"
        >
          0
        </p>
        <p>
          Bound Applications
        </p>
      </div>
    </section>
  </div>
</div>
`;
